src/compiler/commandLineParser.ts

```diff
@@ -389,23 +395,449 @@ module ts {
+                        contains(entries, entry.substr(0, entry.length - 5) + ".ts")) {
+                        continue;
                     }
+                    
+                    // This wildcard has no further directory to process, so include the file.
+                    includeFile(path, /*isLiteralFile*/ false);
                 }
+                else if (host.directoryExists(path)) {
+                    // If this was a directory, process the directory.
+                    if (expandFileSpec(path, fileSpec, offset + 1, end) === ExpandResult.Error) {
+                        return ExpandResult.Error;
+                    }
+                }
+            }
+            
+            return ExpandResult.Ok;
+        }
+        
+        /**
+          * Expands a `**` recursive directory wildcard.
+          * @param basePath The directory to recursively expand.
+          * @param fileSpec The original file specification.
+          * @param start The starting offset in the file specification.
+          * @param end The end offset in the file specification.
+          */
+        function expandRecursiveDirectory(basePath: string, fileSpec: string, start: number, end: number): ExpandResult {
+            if (isExpandingRecursiveDirectory) {
+                if (errors) {
+                    errors.push(createCompilerDiagnostic(Diagnostics.File_specification_cannot_contain_multiple_recursive_directory_wildcards_Asterisk_Asterisk_Colon_0, fileSpec))
+                }
+                return ExpandResult.Error;
             }
-            return files;
+            
+            // expand the non-recursive part of the file specification against the prefix path.
+            isExpandingRecursiveDirectory = true;
+            let result = expandFileSpec(basePath, fileSpec, start, end);
+            isExpandingRecursiveDirectory = false;
+            
+            if (result !== ExpandResult.Error) {
+                // Recursively expand each subdirectory.
+                let entries = host.readDirectoryFlat(basePath);
+                for (let entry of entries) {
+                    let path = combinePaths(basePath, entry);
+                    if (host.directoryExists(path)) {
+                        if (expandRecursiveDirectory(path, fileSpec, start, end) === ExpandResult.Error) {
+                            result = ExpandResult.Error;
+                            break;
+                        }
+                    }
+                }
+            }
+            
+            return result;
+        }
+        
+        /**
+          * Includes a file in a file set.
+          * @param file The file to include.
+          * @param isLiteralFile A value indicating whether the file to be added is 
+          *                      a literal file, or the result of matching a file specification.
+          */
+        function includeFile(file: string, isLiteralFile: boolean): void {

```

Comment:
It's better to add the output array as a function parameter to decouple this function from expandFiles instead using a closure to access output. This increases readability, decoupling, testing and performance of this function. This same comment probably applies to other functions defined in expandFiles too.